Makes the exception names of the argument parser consistent

smehringer opened the pull request:
Resolves #481

I changed the names according to our resolution in the strategy meeting ([card](https://github.com/orgs/seqan/projects/4#card-26600178))
Additionally, the first commit get's rid of the internal exception in the function retrive_value. The success is indicated by a return value instead and the exception is thrown outside. This also made the exceptions std::type_conversion-error and std::overflow_errow obsolete which lowers the number of exceptions in general.

MitraDarja commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -285,84 +292,80 @@ class format_parse : public format_base
+     * \param[out] value Stores the cast value.
+     * \param[in] in The input argument to be cast.
+     * \returns seqan3::option_cast_result::type_conversion_error if \p in could not be converted to an arithmetic type
+     *          via std::from_chars, seqan3::option_cast_result::overflow_error if \p in could be converted but the
```
What does \p stand for?
Do we always shorten \param[in] with \p in?

MitraDarja commented on include/seqan3/argument_parser/exceptions.hpp:
```diff
@@ -19,6 +19,17 @@
 namespace seqan3
 {
 
+//!\brief This class is eprecated.
```
deprecated

smehringer commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -285,84 +292,80 @@ class format_parse : public format_base
-     * \throws seqan3::type_conversion_failed
+     * \param[out] value Stores the cast value.
+     * \param[in] in The input argument to be cast.
+     * \returns seqan3::option_cast_result::type_conversion_error if \p in could not be converted via the stream
```
With \p doxygen just formats the next word in code style. I will change it to back ticks ` which isused more frequently

smehringer commented:
sorry I did not see that you approved already. You can or cannot check the differences and approve again. In general if you find typos, you can request changes as they are just as valid as code changing requests.

rrahn commented on include/seqan3/argument_parser/exceptions.hpp:
```diff
@@ -133,13 +134,13 @@ class validation_failed : public parser_invalid_argument
  * - Both identifiers must not be empty (one is ok)
  */
-class parser_design_error : public std::logic_error
+class design_error : public std::logic_error
```
in the Changelog you mentioned, that this inherits from argument_parser_error, which I would prefer. I also would not inherit the argument_parser_error from invalid_argument but make it it's own exception.

smehringer replied:
> in the Changelog you mentioned, that this inherits from argument_parser_error, which I would prefer.

Good catch

> I also would not inherit the argument_parser_error from invalid_argument but make it it's own exception.

You mean it should inherit from std::exception instead of std::invalid_argument?

rrahn replied:
Yes, because design_error is not an invalid_argument is it? I would think it is an own error for itself.

rrahn commented on include/seqan3/argument_parser/exceptions.hpp:
```diff
@@ -34,93 +45,83 @@ namespace seqan3
 
 //!\brief Argument parser exception thrown when an argument could not be casted to the according type.
-class validation_failed : public parser_invalid_argument
+class validation_error : public argument_parser_error
```
So if I get this right. When people use one of this exceptions they will get an compiler error because it is no longer present. When they used parser_invalid_argument they will get a deprecation message but also thrown exceptions are never catched and the program terminates. So which means we have two options. Either deprecate everything or let the compiler always fail when using old exceptions. I am tending for the former solution. Or am I missing something here?

smehringer replied:
Alright, I will deprecate all the old names and we can remove them with 3.1

rrahn commented on CHANGELOG.md:
```diff
@@ -54,6 +54,13 @@ If possible, provide tooling that performs the changes, e.g. a shell-script.
+  * The exception seqan3::validation_failed was renamed to seqan3::validation_error.
+  * The exception seqan3::parser_design_error was renamed to seqan3::design_error and also inherits from
+    seqan3::argument_parser_error.
+  * The exception seqan3::type_conversion_failed was removed, you can catch seqan3::user_input_error instead.
```
should then be deprecated.

rrahn commented on CHANGELOG.md:
```diff
@@ -54,6 +54,13 @@ If possible, provide tooling that performs the changes, e.g. a shell-script.
+  * The exception seqan3::parser_design_error was renamed to seqan3::design_error and also inherits from
+    seqan3::argument_parser_error.
+  * The exception seqan3::type_conversion_failed was removed, you can catch seqan3::user_input_error instead.
+  * The exception seqan3::overflow_error was removed, you can catch seqan3::user_input_error instead.
```
should then be deprecated.

smehringer commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -285,84 +292,80 @@ class format_parse : public format_base
+        typename container_option_t::value_type tmp{};
 
-        retrieve_value(tmp, in); // throws on failure
+        auto res = cast_option_value(tmp, in); // throws on failure
```
Hm.. So when there is a conversion error the tmp variable will not be altered so I am appending the default constructed value. So this is not an issue but probably it is better to not append anything at all.

smehringer commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -285,84 +292,80 @@ class format_parse : public format_base
-     *
-     * \param[out] value container that stores the casted value.
-     * \param[in]  in    The input argument to be casted.
+    /*!\brief Appends a cast value to its container.
```
I talked briefly with @marehr about this and decided to switch from `cast` (which I agree is not appropiate) to `parse` everywhere in the documentation.

smehringer commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -406,22 +408,58 @@ class format_parse : public format_base
+
+        if (res == option_cast_result::conversion_error)
+        {
+            throw user_input_error{msg + "Argument " + input_value + " could not be cast to type " +
```
see above -> `parsed`

smehringer commented:
Sorry, I had to rebase and add a different commit for the renaming in the documentation to keep the history clean. Doing that later would have been too compilcated.

rrahn commented on include/seqan3/argument_parser/argument_parser.hpp:
```diff
@@ -162,15 +162,15 @@ class argument_parser
     argument_parser & operator=(argument_parser &&) = default;      //!< Defaulted.
 
-    /*!\brief Initializes an argument_parser object from the command line arguments.
+    /*!\brief Initializes seqan3::an argument_parser object from the command line arguments.
```
```suggestion
    /*!\brief Initializes an seqan3::argument_parser object from the command line arguments.
```

rrahn commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -406,22 +413,58 @@ class format_parse : public format_base
+
+        if (res == option_parse_result::error)
+        {
+            throw user_input_error{msg + "Argument " + input_value + " could not be cast to type " +
```
```suggestion
            throw user_input_error{msg + "Argument " + input_value + " could not be parsed as type " +
```

rrahn commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -283,86 +290,87 @@ class format_parse : public format_base
 
-        retrieve_value(tmp, in); // throws on failure
-        value.push_back(tmp);
+        auto res = parse_option_value(tmp, in); // throws on failure
```
```suggestion
        auto res = parse_option_value(tmp, in);
```

rrahn commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -406,22 +413,58 @@ class format_parse : public format_base
+    }
+
+    /*!\brief Tries to parse an input string into boolean value.
+     * \param[in] res A result value of parseing an input string to the respective option value type.
```
```suggestion
     * \param[in] res A result value of parsing an input string to the respective option value type.
```

rrahn commented on include/seqan3/argument_parser/detail/format_parse.hpp:
```diff
@@ -406,22 +413,58 @@ class format_parse : public format_base
+
+    /*!\brief Tries to parse an input string into boolean value.
+     * \param[in] res A result value of parseing an input string to the respective option value type.
+     * \param[in] option_name The name of the option who's input was parsed.
```
```suggestion
     * \param[in] option_name The name of the option whose input was parsed.
```
?

rrahn commented:
Have you verified the code coverage issue?

smehringer commented:
It has nothing to do with me. I'll rebase to current master to see if this fixes it.

smehringer commented:
@rrahn Code coverage only fails because of the filesystem errors that in the file validators were renamed but are not tested. So I think this can be merged.

smehringer commented:
@rrahn I updated the Changelog entry to link this PR.
The code coverage fail is unrelated to this PR, since it records 4 misses that are due to the untested error handling of the `input_directory_validator`s where I changed the exception names so I trigger the misses again. Fixes the code coverage should be part of another PR, I'll make a card.
